crates/oxc_allocator/src/vec2/raw_vec.rs

```diff
@@ -720,6 +719,90 @@ fn capacity_overflow() -> ! {
     panic!("capacity overflow")
 }
 
+fn handle_alloc_error(layout: Layout) -> ! {
+    panic!("encountered allocation error: {:?}", layout)
+}
+
+/// Wrapper around `Layout::from_size_align` that adds debug assertions.
+#[inline]
+fn layout_from_size_align(size: usize, align: usize) -> Result<Layout, AllocError> {
+    Layout::from_size_align(size, align).map_err(|_| AllocError)
+}
+
+/// Returns a pointer suitable for holding data described by
+/// a new layout with `layout`’s alignment and a size given
+/// by `new_size`. To
+/// accomplish this, this may extend or shrink the allocation
+/// referenced by `ptr` to fit the new layout.
+///
+/// If this returns `Ok`, then ownership of the memory block
+/// referenced by `ptr` has been transferred to this
+/// allocator. The memory may or may not have been freed, and
+/// should be considered unusable (unless of course it was
+/// transferred back to the caller again via the return value of
+/// this method).
+///
+/// If this method returns `Err`, then ownership of the memory
+/// block has not been transferred to this allocator, and the
+/// contents of the memory block are unaltered.
+///
+/// # Safety
+///
+/// This function is unsafe because undefined behavior can result
+/// if the caller does not ensure all of the following:
+///
+/// * `ptr` must be currently allocated via this allocator,
+///
+/// * `layout` must *fit* the `ptr` (see above). (The `new_size`
+///   argument need not fit it.)
+///
+/// * `new_size` must be greater than zero.
+///
+/// * `new_size`, when rounded up to the nearest multiple of `layout.align()`,
+///   must not overflow (i.e. the rounded value must be less than `usize::MAX`).
+///
+/// (Extension subtraits might provide more specific bounds on
+/// behavior, e.g. guarantee a sentinel address or a null pointer
+/// in response to a zero-size allocation request.)
+///
+/// # Errors
+///
+/// Returns `Err` only if the new layout
+/// does not meet the allocator's size
+/// and alignment constraints of the allocator, or if reallocation
+/// otherwise fails.
+///
+/// Implementations are encouraged to return `Err` on memory
+/// exhaustion rather than panicking or aborting, but this is not
+/// a strict requirement. (Specifically: it is *legal* to
+/// implement this trait atop an underlying native allocation
+/// library that aborts on memory exhaustion.)
+///
+/// Clients wishing to abort computation in response to a
+/// reallocation error are encouraged to call the [`handle_alloc_error`] function,
+/// rather than directly invoking `panic!` or similar.
+///
+/// [`handle_alloc_error`]: ../../alloc/alloc/fn.handle_alloc_error.html
+unsafe fn realloc(
```

Comment:
One thing is confusing me - where is this function's code copied from?

The doc comment looks like it came from here: https://github.com/fitzgen/bumpalo/blob/1d2fbea9e3d0c2be56367b9ad5382ff33852a188/src/alloc.rs#L322-L402

But the body of the function looks more like (but not exactly the same as) this one: https://github.com/fitzgen/bumpalo/blob/1d2fbea9e3d0c2be56367b9ad5382ff33852a188/src/lib.rs#L1917-L1936

Reply:
I am sorry this confused you. Yes, the comment and the function implementation are copied from these two places.

Reply:
Ah ha OK. Thanks for clarifying.